storage_policies/stack/__main__.py

```diff
@@ -186,3 +186,77 @@ def add_bucket_permissions(name: str,
                            release_access_group,
                            release_bucket,
                            'roles/storage.objectViewer')
+
+# Cloud Run is used for the server component of the analysis-runner.
+cloudrun = gcp.projects.Service('cloudrun-service',
+                                service='run.googleapis.com',
+                                disable_on_destroy=False)
+
+analysis_runner_service_account = gcp.serviceaccount.Account(
+    'analysis-runner-service-account',
+    account_id='analysis-runner-server',
+    display_name='analysis-runner service account',
+    opts=pulumi.resource.ResourceOptions(depends_on=[cloudidentity]))
+
+secretmanager = gcp.projects.Service('secretmanager-service',
+                                     service='secretmanager.googleapis.com',
+                                     disable_on_destroy=False)
+
+# Add a secret that will hold the Hail token.
+hail_token_secret = gcp.secretmanager.Secret(
+    'hail-token-secret',
+    replication=gcp.secretmanager.SecretReplicationArgs(
+        user_managed=gcp.secretmanager.SecretReplicationUserManagedArgs(
+            replicas=[
+                gcp.secretmanager.SecretReplicationUserManagedReplicaArgs(
+                    location=REGION
+                ),
+            ],
+        ),
+    ),
+    secret_id='hail-token',
+    opts=pulumi.resource.ResourceOptions(depends_on=[secretmanager]))
+
+# The analysis-runner server needs to read the Hail token secret.
+gcp.secretmanager.SecretIamMember(
+    'analysis-runner-service-account-secret-reader',
+    secret_id=hail_token_secret.id,
+    role='roles/secretmanager.secretAccessor',
+    member=pulumi.Output.concat('serviceAccount:', analysis_runner_service_account.email))
+
+project_number = gcp.organizations.get_project().number
+
+# Allow the Cloud Run Service Agent to pull the image.
+repo_member = gcp.artifactregistry.RepositoryIamMember(
+    'artifact-registry-reader',
+    project='analysis-runner',
+    location=REGION,
+    repository='images',
+    role='roles/artifactregistry.reader',
+    member=f'serviceAccount:service-{project_number}@serverless-robot-prod.iam.gserviceaccount.com')
+
+analysis_runner_server = gcp.cloudrun.Service(
+    'analysis-runner-server',
+    location=REGION,
+    autogenerate_revision_name=True,
+    template=gcp.cloudrun.ServiceTemplateArgs(
+        spec=gcp.cloudrun.ServiceTemplateSpecArgs(
+            containers=[gcp.cloudrun.ServiceTemplateSpecContainerArgs(
+                envs=[{'name': 'DATASET', 'value': dataset}],
+                image=f'australia-southeast1-docker.pkg.dev/analysis-runner/images/server:06dd37129625'
```

Comment:
Wondering where is the server image built?

And where is the tag `06dd37129625` coming from - is it a commit hash? Should it be hardcoded or parameterized? Do you see it to be updated often?

Reply:
Sorry, this is where it would have been helpful to send the "server" code for review first :).

At the moment the image is just built by hand, but we could easily add some automation to it (I hope we won't have to rebuild it often at all though).

The reason the version is hardcoded here is that I'd like the repository to represent the "current state in production". However, there's also an argument that we could just use "latest" here and always have Pulumi update the image when it's rerun. I'm actually not entirely sure whether Pulumi would resolve "latest" and then check whether the underlying image has changed or not -- do you know?

We'll have a similar situation for the server image, which currently is based on the driver image ;). We could either hardcode the base version for the server image or use "latest".

What do you prefer?

Reply:
You are right that Pulumi doesn't know whether the image sitting under the `latest` has changed. It only updates the stack if the tag string has changed. 

It might make sense to do that for Pulumi: add some code into `__main__.py` that would check the tag of the most recent container. We can even version images semantically (`v1.0.0`) so it's easier to clearly distinguish the latest stable one. And then substitute the tag in the code, so Pulumi triggers only on new tags:

```
version = <...>
analysis_runner_server = gcp.cloudrun.Service(
<...>
                image=f'australia-southeast1-docker.pkg.dev/vlad-dev/cpg-qc/cloudrun-helloworld:{version}'
<...>
```

What do you think? Might be unnecessary if the images won't change that often.

Reply:
I tried to get this working using the `pulumi_docker` module, but couldn't get the authentication to work (see my message on Slack). Happy to revise that if you have more ideas though.

FWIW I found this article pretty helpful to understand how Terraform handles this problem in general:
https://medium.com/faun/understanding-terraform-latest-in-docker-image-image-id-latest-33ca1bebe66b

This doesn't translate directly to Pulumi though as far as I understand.

Reply:
Thanks for the pointer, and again for exploring options! All good for this PR - okay for merge from me when you are ready.